File: Heap&Priority Queue/2. Last Stone Weight/solution.go
func lastStoneWeight(stones []int) int {
	sort.Slice(stones, func(i, j int) bool {
		return stones[i] > stones[j]
	})
	return recursive(stones)
}

func recursive(stones []int) int {
	if len(stones) == 1 {
		return stones[0]
	} else if len(stones) == 0 {
		return 0
	} else {
		slice := stones[:]
		subset := slice[:2]
		tail := slice[2:]
		diff := subset[0] - subset[1]
		if diff == 0 {
			return recursive(tail)
		} else {
			newStones := append(tail, diff)
			sort.Slice(newStones, func(i, j int) bool {
				return newStones[i] > newStones[j]
			})
			return recursive(newStones)
		}
	}
}

File: Heap&Priority Queue/2. Last Stone Weight/solution.exs
defmodule Solution do
  @spec last_stone_weight(stones :: [integer]) :: integer
  def last_stone_weight(stones) do
    recursive(Enum.sort(stones, :desc))
  end

  def recursive([result]), do: result
  def recursive([]), do: 0
  def recursive([xl, l | tail]) when xl == l do
    recursive(tail)
  end

  def recursive([xl, l | tail]) do
    recursive([xl - l | tail] |> Enum.sort(:desc))
  end
end
